In [1]:
import numpy as np
import numericalMethods as nm


# **Numerical Analysis Math 4610**
### **Test 2**
- **Please write clearly and neatly.**
- **Show all the required work, explaining necessary steps.**

1. 
(a) Use appropriate Lagrange interpolating polynomials of degrees one and two to approximate the following: $f(0.25)$ if $f(−1) = 0.86199480, f(−0.5) = 0.95802009, f(0) = 1.0986123, f(0.5) = 1.2943767$.

Using thm. 3.2, If $x_0, x_1, ..., x_n$ are n + 1 distinct numbers and $f$ is a function whose values are given at these numbers, then a unique polynomial $P(x)$ of degree at most n exists such that $P(x_i) = f(x_i)$ for each $i = 0, 1, ..., n$ given by: $$P(x) = \sum_{i=0}^n f(x_i) \frac{L_i(x)}{L_i(x_i)}$$ where $$L_i(x) = \prod_{j=0, j \neq i}^n \frac{x - x_j}{x_i - x_j}$$

For degree 1, because the estimate is $f(0.25)$, $L_0$ and $L_1$ will use the points at $f(0)$ and $f(0.5)$

$$L_0(x) = f(x_0)\frac{x-x_1}{x_0-x_1} = f(x_0)\frac{x-0.5}{-0.5}$$ 
$$L_1(x) = f(x_1)\frac{x-x_0}{x_1-x_0} = f(x_1)\frac{x}{0.5}$$

$$P(x) = 1.0986123\frac{x-0.5}{-0.5} + 1.2943767\frac{x}{0.5}$$
$$P(0.25) = 1.0986123\frac{0.25-0.5}{-0.5} + 1.2943767\frac{0.25}{0.5} = 1.1964945$$

For degree 2, because the estimate is $f(0.25)$, $L_0$, $L_1$, and $L_2$ will use the points at $f(-0.5)$, $f(0)$, and $f(0.5)$

$$L_0(x) = f(x_0)\frac{(x-x_1)(x-x_2)}{(x_0-x_1)(x_0-x_2)} = f(x_0)\frac{(x-0)(x-0.5)}{0.5}$$

$$L_1(x) = f(x_1)\frac{(x-x_0)(x-x_2)}{(x_1-x_0)(x_1-x_2)} = f(x_1)\frac{(x+0.5)(x-0.5)}{-0.25}$$

$$L_2(x) = f(x_2)\frac{(x-x_0)(x-x_1)}{(x_2-x_0)(x_2-x_1)} = f(x_2)\frac{(x+0.5)(x)}{0.5}$$

$$P(x) = 0.95802009\frac{(x+0.5)(x)}{0.5} + 1.0986123\frac{(x-0)(x-0.5)}{-0.25} + 1.2943767\frac{(x+0.5)(x-0.5)}{-0.25}$$

$$P(0.25) = 1.18959797625$$

$\\$

(b) The above data were generated using the function $y = ln(e^x + 2)$. Use the error formula to find a bound for the error, and compare the bound to the actual error for the cases $n = 1$ and $n = 2$.

Using thm. 3.3 since $x_0,x_1,...,x_n$ are distinct numbers in the interval $[0, 0.5]$ and $[-0.5, 0.5]$ and $f \in C^2$ and $C^3$ on their intervals. Then, for each x in the interval, a number $\xi(x)$ between the min and max exists with:

$$ E(x) = \frac{f^{n+1} \xi(x) }{(n+1)!}(x-x_0)(x-x_1)...(x-x_n) $$
$n = 1$


$$f^2(x) = \frac{2e^x}{(e^x+2)^2}$$
The maximum value on the interval for $\xi(x)$ $[0,0.5]$ is at $0.5$
The maximum value on the interval for $g(x) = x(x-0.5)$ is at $-0.25$
$$ \implies \frac{f^2(\xi(x))}{2}(x-x_0)(x-x_1)$$
$$\implies \frac{f^2(\xi)}{2}(x-0)(x-0.5) \le |0.247682806306||-0.0625| = 0.0154801753941$$

$n = 2$
$$f^3(x) = \frac{2e^x(-e^x+2)}{(e^x+2)^3}$$

The maximum value on the interval for $\xi(x)$, $[-0.5,0.5]$ is at $-0.5$
The maximum value on the interval for $g(x) = x(x-0.5)(x+0.5)$, $[-0.5,0.5]$ is at $-0.2887$ and $0.2887$ both with an absolute value of $0.0481 $

$$\frac{f^{3} \xi(x) }{6}(x+0.5)(x)(x-0.5)$$

$$\implies \frac{f^3(\xi)}{6}(x+0.5)(x)(x-0.5) \le |0.0481||0.0955| = 0.00459247923792$$

The actual error is equal to $p-p^*$

For degree 1 : $1.18906993111 - 1.1964945$ = $-0.00742456889$ the actual error $0.00742456889$

For degree 2 : $1.18959797625 - 1.1964945$ = $-0.00689652375$ the actual $0.00689652375$

-----

2.
(a) Let $f(x) = 3xe^x−e^{2x}$. Approximate $f(1.03)$ by the Hermite interpolating polynomial of degree at most three using $x_0 = 1$ and $x_1 = 1.05$.

In [6]:
def f(x):
    return 3*x*np.exp(x)-np.exp(2*x)
x = np.array([1,1.05])
y = np.array([f(x[0]),f(x[1])])
dy = np.array([1.53157877289,1.24221455095])
nm.HermiteInterpolation(x,y,dy)

array([[ 1.        ,  0.76578939,  0.        ,  0.        ,  0.        ],
       [ 1.        ,  0.76578939,  1.53157877,  0.        ,  0.        ],
       [ 1.05      ,  0.83543111,  1.39283446, -2.7748863 ,  0.        ],
       [ 1.05      ,  0.83543111,  1.24221455, -3.01239813, -4.75023662]])

Using thm. 3.9 since $f \in c^1$ $[1,1.05]$ and $x_0,x_1$ are distinct numbers in the interval $[1,1.05]$, the distinct unique polynomial of least degree with $f$ and $f'$ at $x0$ and $x1$ is the Hermite polynomial of degree at most $2(1)+1 = 3$ is given by:
$$H_3(x) = f(x_0) + \sum_{k=1}^{n=2} f(x_0,x_1,...,x_k)(x-x_0)...(x-x_{k-1})$$

$$H_3(x) = 0.76578939 + 1.53157877(x-1) - 2.7748863(x-1)^2 - 4.75023662(x-1)^2(x-1.05)$$

$$H_3(1.03) = 0.76578939 + 1.53157877(0.03) - 2.7748863(0.03)^2 - 4.75023662(0.03)^2(-0.02)$$

$$H_3(1.03) = 0.809324859689$$


$\\$

(b) Compare the actual error to the error bound.

The actual error is given by $p-p^*$
$$H_3(1.03) - f(1.03) = 0.809324859689 - 0.809323618902 = 0.00000124078729324$$ 
The absolute error is $0.00000124078729324$



-----

3. (No Computer Programming)

(a) Construct the natural cubic spline for the following data.

$$ \begin{array}{c|c}
 \text{x} & \text{f(x)} \\ \hline \\
 \text{-0.25 } & \text{1.33203} \\ \hline \\
 \text{0.25} & \text{0.80078} \\ 
 \end{array} $$

Using 3.11 because $f$ is defined at $a = x_0 < x_1 < ... < x_n = b, f$ has a unique natural spline interpolant $S$ on the nods $x_0, x_1, ..., x_n$. The natural spline interpolant is given by:


$$S_i = a_i + b_i(x-x_i) + c_i(x-x_i)^2 + d_i(x-x_i)^3$$
For $$x_i \le x \le x_{i+1}$$

Points:
- $(-0.25, 1.33203), (0.25, 0.80078)$

$$S_0 = a_0 + b_0(x+0.25) + c_0(x+0.25)^2 + d_0(x+0.25)^3$$

$$h_n = x_{n+1} -x_n$$

$$h_0 = 0.5$$

<br>

$$c_n = A^{-1}u$$

$$A = \begin{bmatrix} 1 & 0  \\ 0 & 1  \\ \end{bmatrix}$$
$$b = \begin{bmatrix} 0  \\ 0 \\ \end{bmatrix}$$

$$Ac = u$$
$$c_n = [0,0]$$

<br>

$$b_0 = \frac{1}{h_0}(a_1-a_0) - \frac{h_0}{3}(c_1+2c_0)$$
$$b_0 = \frac{1}{0.5}(0.80078-1.33203) - (0) = -1.0625$$
$$d_0 = \frac{1}{3h_0}(c_1-c_0) = 0$$

<br>

$$ \implies S(x) = S_0(x) = 1.33203 - 1.0625(x+0.25) + 0(x+0.25)^2 + 0(x+0.25)^3$$

$$ \implies S(x) = 1.33203 - 1.0625(x+0.25)$$ 
For $-0.25 \le x \le 0.25$



$\\$

(b) Construct the clamped cubic  spline using the data above and the fact that $\\$ $f′(−0.25)=0.437500$ and $f′(0.25)=−0.625000$

Because  $f$ is defined at a = -0.25 < 0.25 and is differentiable at $a$ and $b$, $f$ has a unique clamped spine interpolated at $-0.25$ and $0.25$ that satisfies $s'(-0.25) = f'(-0.25)$ and $s'(0.25) = f'(0.25)$

Given by:

$$S_i = a_i + b_i(x-x_i) + c_i(x-x_i)^2 + d_i(x-x_i)^3$$
For $$x_i \le x \le x_{i+1}$$


Points:
- $(-0.25, 1.33203), (0.25, 0.80078)$

$$S_0 = a_0 + b_0(x+0.25) + c_0(x+0.25)^2 + d_0(x+0.25)^3$$

$$ h_n = x_{n+1} -x_n$$
<br>
$$h_0 = 0.5$$

$$ \\ A = \begin{bmatrix} 2h_0 & h_0 \\ h_0 & 2h_0 \\ \end{bmatrix}$$
$$A = \begin{bmatrix} 1 & 0.5 \\ 0.5 & 1 \\ \end{bmatrix}$$

$$u = \begin{bmatrix} \frac{3}{h_0}(a_1-a_0) -3f'(a)\\ 3f'(b) -\frac{3}{h_0}(a_1-a_2)\\ \end{bmatrix}$$

$$u = \begin{bmatrix} \frac{3}{0.5}(0.80078-1.33203) -3(0.4375)\\ 3(-0.625) -\frac{3}{0.5}(0.80078-1.33203)\\ \end{bmatrix}$$
$$u = \begin{bmatrix} −4.5\\ 1.3125\\ \end{bmatrix}$$

$$Ac = u$$

$$\begin{bmatrix} 1 & 0.5 \\ 0.5 & 1 \\ \end{bmatrix} \begin{bmatrix} c_0\\ c_1\\ \end{bmatrix} = \begin{bmatrix} −4.5\\ 1.3125\\ \end{bmatrix}$$

$$c = [-6.875, 4.75]$$

<br>

$$ b_0 = \frac{1}{h_0}(a_1-a_0) - \frac{h_0}{3}(c_1+2c_0)$$
$$b_0 = \frac{1}{0.5}(0.80078-1.33203) - \frac{(0.5)}{3}(4.75+2(-6.875)) = 0.4375$$

<br>

$$ \\ d_0 = \frac{1}{3h_0}(c_1-c_0) = \frac{1}{3(0.5)}(4.75-(-6.875)) = 7.75$$

<br>

$$ \\ \implies S(x) = S_0(x) = 1.33203 - 0.4375(x+0.25) -6.875 (x+0.25)^2 + 7.75(x+0.25)^3, -0.25 \le x \le 0.25$$


-----

4.
(a) Use the forward-difference formulas and backward-difference formulas to determine each missing entry in the following tables.

$$ \begin{array}{c|c|c}
 \text{x} & \text{f(x)} & \text{f'(x)} \\ \hline \\
 \text{1.0 } & \text{1.000} \\ \hline \\
 \text{1.2 } & \text{1.2625} \\ \hline \\
 \text{1.4} & \text{1.6595} \\ 
 \end{array} $$

Forward Divided Difference:
the forward difference is given by:
$$ \implies f'(x_0) = \frac{f(x_1)-f(x_0)}{x_1-x_0} = 1.3125$$
$$ \implies f'(1) = \frac{1.2625-1}{.2}$$
Backward Divided Difference
the backward difference is given by:
$$ \implies f'(x_2) = \frac{f(x_2)-f(x_1)}{x_2-x_1}$$
$$ \implies f'(1.4) = \frac{1.6595-1.2625}{1.4-1.2} = 1.985$$
Final Divided Difference
$$ \implies f'(x_1) = \frac{f'(x_2)-f'(x_0)}{x_2-x_0}$$
$$ \implies f'(1.2) = \frac{1.6595-1.2625}{1.4-1} = 1.68125$$

The final table is:
$$ \begin{array}{c|c|c}
 \text{x} & \text{f(x)} & \text{f'(x)} \\ \hline \\
 \text{1.0 } & \text{1.000} & \text{1.3125} \\ \hline \\
 \text{1.2 } & \text{1.2625} & \text{1.68125} \\ \hline \\
 \text{1.4} & \text{1.6595} & \text{1.985} \\
 \end{array} $$

$$\\$$

(b) The above data is from the function $f (x) = x^2 ln x + 1$. Compute the actual errors and find the error bounds using the error formulas.

The derivative of $f(x) = x^2 ln x + 1$ is $f'(x) = 2xlnx + x^2\frac{1}{x} + 0 = 2xlnx + x^2$

The actual error of $f'(1) = p-p* = 1 - 1.3125 = -0.3125$, the absolute error is $0.3125$

The actual error of $f'(1.2) = p-p* = 1.2625 - 1.68125 = -0.41875$, the absolute error is $0.41875$

The actual error of $f'(1.4) = p-p* = 1.6595 - 1.985 = -0.3255$, the absolute error is $0.3255$

-----

5. 
Determine the values of n and h required to approximate $\int_0^2 x^2sin x \,dx$ to within $10^{−4}.\\$
Use
(a) Composite trapezoidal rule

$\\$

Using thm. 4.5 Because $f \in C^2[a,b], h=\frac{(b-a)}{n}$ and $x_j = a+jh$ there exists a $\mu \in (a,b)$ for which the composite trapozoidal rule error term can be given by:

$$\frac{b-a}{12}h^2f''(\mu)$$

Allowing to find $h$ then $n$ since $h = \frac{b-a}{n}$

To find h for the composite trapezoidal rule:
$$ \frac{b-a}{12}h^2f^2(\mu) \le 10^{-4} $$
$$ f^2(x) = -x^2sin(x)+4xcos(x)+2sin(x) $$
The maximum of the error is at $x = 2$ with $|f(x)| = 5.418$

$$ \implies \frac{2-0}{12}h^2f^2(\mu) \le 10^{-4} $$

$$ \implies \frac{1}{6}h^2f^2(\mu) \le 10^{-4} $$

$$ \implies h^2(5.418) \le 6*10^{-4} $$

$$ \implies h \le \sqrt{\frac{6*10^{-4}}{5.418}} $$

$$ \implies h \le 0.0105234011235 $$

To find n for the composite trapezoidal rule:

$$ \frac{b-a}{h} = n $$

$$ \implies \frac{2-0}{0.0105234011235} \ge n $$

$$ \implies n \ge 190.05 $$

In [33]:
def f(x):
    return np.power(x,2)*np.sin(x)

nm.composite_trapezoidal(f,0,2,191)

2.4695014047090926

$\\$

(b) Composite Simpson’s rule

Using 4.4 because $f \in C^4[a,b]$, $n$ be even, $h=\frac{(b-a)}{(n)}$ and $x_j a+jh$ there exists a $\mu \in (a,b)$ for which the composite simpson rule error term can be given by:

$$\frac{b-a}{180}h^4f^4(\mu)$$

allowing to find $h$ then $n$ since $h = \frac{b-a}{n}$

to find h for the composite Simpson's rule:
$$ \frac{b-a}{180}h^4f^4(\mu) \le 10^{-4} $$
$$ f^4(x) = -x^2sin(x)+8xcos(x)+12sin(x) $$
The maximum of the error is at $x = 1.059$ with $|f(x)| = 13.634$

$$ \implies \frac{2-0}{180}h^4f^4(\mu) \le 10^{-4} $$
$$ \implies \frac{1}{90}h^4f^4(\mu) \le 10^{-4} $$
$$ \implies h^4(13.634) \le 10^{-4} * 90 $$
$$ \implies h \le \sqrt[4]{\frac{10^{-4} * 90}{13.634}} $$
$$ \implies h \le 0.160289400628 $$

to find n for the composite Simpson's rule:

$$ \frac{a-b}{h} = n $$
$$ \implies \frac{2-0}{n} \ge 0.160289400628 $$
$$ \implies \frac{2-0}{0.160289400628} \ge n $$
$$ \implies n \ge 12.477431 $$
Because n must be even for Simpson's rule, round up to 14.

In [9]:
nm.composite_simpson(f,0,2,14)

2.469442070644263

$\\$

(c) Composite Midpoint rule

Using thm 4.6, because $f \in C^2[a,b]$, $n$ be even, $h=\frac{(b-a)}{(n+2)}$ and $x_j a+(j+1)h$ there exists a $\mu \in (a,b)$ for which the composite trapozoidal rule error term can be given by:

$$\frac{b-a}{6}h^2f''(\mu)$$

allowing to find $h$ then $n$ since $h = \frac{b-a}{n+2}$

to find h for the composite Midpoint rule:
$$ \frac{b-a}{6}h^2f^2(\mu) \le 10^{-4} $$
$$ f^2(x) = -x^2sin(x)+4xcos(x)+2sin(x) $$
The maximum of the error is at $x = 2$ with $|f(x)| = 5.418$

$$ \implies \frac{2-0}{6}h^2f^2(\mu) \le 10^{-4} $$
$$ \implies \frac{1}{3}h^2f^2(\mu) \le 10^{-4} $$
$$ \implies h^2(5.418) \le 10^{-4} * 3 $$
$$ \implies h \le \sqrt{\frac{10^{-4} * 3}{5.418}} $$
$$ \implies h \le 0.00744116829561 $$

to find n for the composite Midpoint rule:

$$ \frac{a-b}{n+2} = h $$
$$ \implies \frac{2-0}{n+2} \ge 0.00744116829561 $$
$$ \implies \frac{2}{0.00744116829561} \ge n+2 $$
$$ \implies n \ge 266.774998837 $$

because n must be even for the composite midpoint rule, round up to 268.

In [45]:
nm.composite_midpoint(f,0,2,268)

2.45592093188472

-----

6. Use adaptive quadrature to find approximations to within $10^{−4}$ for the following integrals. $\\ \int_0^2 x^2sin x \,dx$

In [20]:
nm.AdaptiveQuadrature(f,0,2,1e-4)

2.4694837480191314

-----